java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java

```diff
@@ -86,7 +88,38 @@
     private final KinesisProducerConfiguration config;
     private final Map<String, String> env;
     private final AtomicLong messageNumber = new AtomicLong(1);
-    private final Map<Long, SettableFuture<?>> futures = new ConcurrentHashMap<>();
+    private final Map<Long, SettableFutureTracker> futures = new ConcurrentHashMap<>();
+    private final PriorityQueue<SettableFutureTracker> oldestFutureTrackerHeap = new PriorityQueue(new SettableFutureTrackerComparator());
```

Comment:
Note that this implementation is not synchronized. Multiple threads should not access a PriorityQueue instance concurrently if any of the threads modifies the queue. Instead, use the thread-safe PriorityBlockingQueue class.

Reply:
Changed